lib/dialects/abstract/connection-manager.js

```diff
@@ -248,64 +217,80 @@ class ConnectionManager {
+            _options.logging = () => {};
+            _options.logging.__testLoggingFn = true;
+
+            return this.sequelize.databaseVersion(_options).then(version => {
+              this.sequelize.options.databaseVersion = semver.valid(version) ? version : this.defaultVersion;
+              this.versionPromise = null;
+
+              return this._disconnect(connection);
+            });
+          }).catch(err => {
+            this.versionPromise = null;
+            throw err;
           });
-        }).catch(err => {
-          this.versionPromise = null;
-          throw err;
-        });
       }
     } else {
       promise = Promise.resolve();
     }
 
     return promise.then(() => {
-      return Promise.race([
-        this.pool.acquire(options.priority, options.type, options.useMaster),
-        new Promise((resolve, reject) =>
-          timers.setTimeout(() => {
-            if (this.poolError) {
-              reject(this.poolError);
-            }
-          }, 0))
-      ])
-        .tap(() => { debug('connection acquired'); })
-        .catch(e => {
-          e = this.poolError || e;
-          this.poolError = null;
-          throw e;
-        });
+      return this.pool.acquire(options.priority, options.type, options.useMaster)
+        .then(mayBeConnection => this._determineConnection(mayBeConnection))
+        .tap(() => { debug('connection acquired'); });
     });
   }
 
   releaseConnection(connection) {
-    return this.pool.release(connection).tap(() => {
-      debug('connection released');
-    });
+    return this.pool.release(connection)
+      .tap(() => { debug('connection released'); })
+      .catch(/Resource not currently part of this pool/, () => {});
+  }
+
+  /**
+   * Check if something acquired by pool is indeed a connection but not an Error instance
+   * Why we need to do this https://github.com/sequelize/sequelize/pull/8330
+   *
+   * @param {Object|Error} mayBeConnection Object which can be either connection or error
+   */
+  _determineConnection(mayBeConnection) {
+    if (mayBeConnection instanceof Error) {
+      return Promise.resolve(this.pool.destroy(mayBeConnection))
+        .catch(/Resource not currently part of this pool/, () => {})
+        .then(() => { throw mayBeConnection; });
+    }
+
+    return Promise.resolve(mayBeConnection);
   }
 
   _connect(config) {
     return this.sequelize.runHooks('beforeConnect', config)
       .then(() => this.dialect.connectionManager.connect(config))
-      .then(connection => this.sequelize.runHooks('afterConnect', connection, config).return(connection));
+      .then(connection => this.sequelize.runHooks('afterConnect', connection, config).return(connection))
+      .catch(error => { return error; });
   }
 
   _disconnect(connection) {
+    // hacky, _determineConnection could send Error instance for disconnection
+    if (connection instanceof Error) {
+      return Promise.resolve();
+    }
```

Comment:
As I mentioned above, I think the hack shouldn't be here.

Delete call of <code>catch</code>.